File: 5-Architecture design patterns/MVP/Day5_Threads/app/src/main/java/com/example/day5_threads/model/Remote/ProductClient.java
package com.example.day5_threads.model.Remote;

import com.example.day5_threads.model.Pojo.ProductResponse;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

public class ProductClient
{


    private static final String BASE_URL = "https://dummyjson.com/";

    private static ProductClient client = null;

    ProductService productService;

    private ProductClient()
    {
        Retrofit retrofit = new Retrofit.Builder()
                .baseUrl(BASE_URL)
                .addConverterFactory(GsonConverterFactory.create())
                .build();

        productService= retrofit.create(ProductService.class);
    }

    public static ProductClient getInstance()
    {
        if(client == null)
        {
            client = new ProductClient();
        }
        return  client;
    }

    public void makeNetworkCall(NetworkCallBack networkCallBack)
    {
        productService.getProducts().enqueue(new Callback<ProductResponse>() {
            @Override
            public void onResponse(Call<ProductResponse> call, Response<ProductResponse> response) {
                if(response.isSuccessful() && response.body() != null)
                {
                    networkCallBack.OnSuccessResult(response.body().getProducts());

                }
                else
                {
                    networkCallBack.OnFailuerResult(response.message());

                }
            }

            @Override
            public void onFailure(Call<ProductResponse> call, Throwable t)
            {
                networkCallBack.OnFailuerResult(t.getMessage());


            }
        });
    }


}

File: 4-Android with Kotlin/Day4/app/src/main/java/com/example/day1/model/ProductClient.kt
package com.example.day1.model

import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory


class ProductClient
private constructor() {
    var productService: ProductService

    init {
        val retrofit = Retrofit.Builder()
            .baseUrl(BASE_URL)
            .addConverterFactory(GsonConverterFactory.create())
            .build()

        productService = retrofit.create<ProductService>(ProductService::class.java)
    }

    fun makeNetworkCall() :List<Product>?
    {
       return productService.getProducts()?.execute()?.body()?.products

    }


    companion object {
        private const val BASE_URL = "https://dummyjson.com/"

        private var client: ProductClient? = null

        val instance: ProductClient?
            get() {
                if (client == null) {
                    client = ProductClient()
                }
                return client
            }
    }
}